### PR TITLE
ONNXRuntime backend

### PR DESCRIPTION
**Pull request by lantiga**
This PR adds the ONNXRuntime (https://github.com/microsoft/onnxruntime) backend, allowing RedisAI to execute deep learning models in ONNX format as well as other machine learning models in ONNX-ML format. It addresses #33.

These latter models can be easily exported from scikit-learn using sklearn-onnx, see https://github.com/onnx/sklearn-onnx.

The PR uses ONNXRuntime version 0.4.0, which only supports loading a model given the filename. This forces us to write the model to a tmp file and load it back to set it in ONNXRuntime. The possibility to load it from an in-memory buffer has been added in `master`: as soon as a new version is released, we'll remove the temp file hack.

Another thing: the buffer containing the model is now cached alongside with the model itself. This is because ONNXRuntime has currently no possibility to re-serialize a model from memory to a buffer, so we need to keep the buffer around. This could change in the future, and in any case it's related to #122.

### CODE REVIEW

File: src/backends/onnxruntime.c [hunk 1]
```diff
@@ -206,8 +237,8 @@ RAI_Model *RAI_ModelCreateORT(RAI_Backend backend, RAI_Device device,
   // loading a model from a buffer, we'll keep this as is
   char tmpfile[256];
   FILE *fp;
-  snprintf(tmpfile,256,"temp-%d.onnx", (int) getpid());
-  fp = fopen(tmpfile,"w");
+  snprintf(tmpfile, 256, "temp-%d.onnx", (int) getpid());
```
Comment (rafie) [inline src/backends/onnxruntime.c]: why not use mkstemp?

File: src/backends/onnxruntime.c [hunk 2]
```diff
@@ -69,11 +69,17 @@ DLDataType RAI_GetDLDataTypeFromORT(ONNXTensorElementDataType dtype) {
   return (DLDataType){ .bits = 0 };
 }
 
-OrtValue* RAI_OrtValueFromTensor(RAI_Tensor* t) {
+OrtValue* RAI_OrtValueFromTensor(RAI_Tensor* t, RAI_Error *error) {
   // TODO: create outside and pass?
   OrtAllocatorInfo* allocator_info;
   OrtStatus* status;
   status = OrtCreateCpuAllocatorInfo(OrtArenaAllocator, OrtMemTypeDefault, &allocator_info);
+  if (status != NULL) {
```
Comment (rafie) [inline src/backends/onnxruntime.c]: It's possible to handle error with a "goto error" to an "error" label at the end of the function, instead of repeating the error block for every failure.

File: src/backends/onnxruntime.c [hunk 3]
```diff
@@ -348,28 +415,41 @@ int RAI_ModelRunORT(RAI_ModelRunCtx* mctx, RAI_Error *error) {
...
+      printf("ERR: non-tensor output from ONNX models, ignoring (currently unsupported).\n");
+    }
     OrtReleaseValue(outputs[i]);
   }
 
   for (size_t i=0; i<n_input_nodes; i++) {
     OrtReleaseValue(inputs[i]);
   }
 
+  OrtReleaseAllocator(allocator);
```
Comment (rafie) [inline src/backends/onnxruntime.c]: Need to call OrtReleaseStatus() is successful?

File: src/backends/onnxruntime.c [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+      case 64:
+        return TF_INT64; break;
+      default:
+        return 0;
+    }
+  }
+  else if (dtype.code == kDLUInt) {
+    switch (dtype.bits) {
+      case 8:
+        return TF_UINT8; break;
```
Comment (rafie) [inline src/backends/onnxruntime.c]: Possible to remove break.

File: src/backends/onnxruntime.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+      case 16:
+        return TF_UINT16; break;
+      default:
+        return 0;
+    }
+  }
+  return 0;
```
Comment (rafie) [inline src/backends/onnxruntime.c]: Possible to remove return.

File: src/backends/onnxruntime.c [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+  // TODO: take from
+  // https://github.com/microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/C_Api_Sample.cpp
+
+  OrtStatus* onnx_status;
+
+  OrtEnv* env;
+  onnx_status = OrtCreateEnv(ORT_LOGGING_LEVEL_WARNING, "test", &env);
+
+  if (onnx_status != NULL) {
```
Comment (rafie) [inline src/backends/onnxruntime.c]: if (onnx_status != NULL || !env) {

File: src/model.c [hunk 7]
```diff
@@ -205,6 +212,9 @@ void RAI_ModelFree(RAI_Model* model, RAI_Error* err) {
   else if (model->backend == RAI_BACKEND_TORCH) {
     RAI_ModelFreeTorch(model, err);
   }
+  else if (model->backend == RAI_BACKEND_ONNXRUNTIME) {
+    RAI_ModelFreeORT(model, err);
+  }
   else {
     RAI_SetError(err, RAI_EUNSUPPORTEDBACKEND, "Unsupported backend\n");
     assert(0);
```
Comment (rafie) [inline src/model.c]: better to abort() as assert() does nothing if NDEBUG is defined.

File: src/backends/onnxruntime.c (hunk 1 above)
Comment (lantiga) [reply src/backends/onnxruntime.c]: Since the command is executed on the main thread, I'm trying to avoid having too many temp files around. BTW, as soon as the new release for ONNXRuntime is out, we can remove this section of the code. Hopefully it'll be before GA

File: src/backends/onnxruntime.c (hunk 5 above)
Comment (lantiga) [reply src/backends/onnxruntime.c]: I'm not sure why here. While it's true that `DLDataTypeCode` can only be `kDLInt`, `kDLInt`, `kDLInt`, the `DLDataType` struct has `code` as `uint8_t` (see https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h#L99), so we can't exclude there's an unexpected value there.

Would you rather `abort`?

File: src/backends/onnxruntime.c (hunk 3 above)
Comment (lantiga) [reply src/backends/onnxruntime.c]: In the ONNXRuntime API, `status` is only ever non-NULL if there's an issue with the function returning `status`. So if everything goes well, `status` is never allocated.

File: src/backends/onnxruntime.c (hunk 2 above)
Comment (lantiga) [reply src/backends/onnxruntime.c]: Done, I just had to enclose code where variable length arrays were initialized in braces to avoid the compiler complaining about goto jumping over variable initialization.

File: src/model.c (hunk 7 above)
Comment (lantiga) [reply src/model.c]: > Also, need to check pointer arguments for null values.

I've added a check for `session` that was missing. Can you maybe suggest other spots?
